chirho/dynamical/internals/ODE/backends/torchdiffeq.py

```diff
@@ -106,75 +107,60 @@ def torchdiffeq_ode_simulate(
     solver: TorchDiffEq,
     dynamics: ODEDynamics,
     initial_state: State[torch.Tensor],
-    timespan,
-):
-    return _torchdiffeq_ode_simulate_inner(
+    start_time: torch.Tensor,
+    end_time: torch.Tensor,
+) -> State[torch.Tensor]:
+    timespan = torch.stack((start_time, end_time))
+    trajectory = _torchdiffeq_ode_simulate_inner(
         dynamics, initial_state, timespan, **solver.odeint_kwargs
     )
+    return trajectory[..., -1].to_state()
 
 
-@ode_simulate_to_interruption.register(TorchDiffEq)
-def torchdiffeq_ode_simulate_to_interruption(
+@_ode_simulate_trajectory.register(TorchDiffEq)
+def torchdiffeq_ode_simulate_trajectory(
     solver: TorchDiffEq,
     dynamics: ODEDynamics,
+    initial_state: State[torch.Tensor],
+    timespan: torch.Tensor,
+) -> State[torch.Tensor]:
+    return _torchdiffeq_ode_simulate_inner(
+        dynamics, initial_state, timespan, **solver.odeint_kwargs
+    )
+
+
+@_ode_get_next_interruptions_dynamic.register(TorchDiffEq)
+def torchdiffeq_get_next_interruptions_dynamic(
+    solver: TorchDiffEq,
+    dynamics: ODEDynamics[torch.Tensor, torch.Tensor],
     start_state: State[torch.Tensor],
-    timespan,  # The first element of timespan is assumed to be the starting time.
-    *,
-    next_static_interruption: Optional["PointInterruption"] = None,
-    dynamic_interruptions: Optional[List["DynamicInterruption"]] = None,
+    start_time: torch.Tensor,
+    next_static_interruption: StaticInterruption,
+    dynamic_interruptions: List[DynamicInterruption],
     **kwargs,
-) -> Tuple[
-    Trajectory[torch.Tensor],
-    Tuple["Interruption", ...],
-    torch.Tensor,
-    State[torch.Tensor],
-]:
-    nodyn = dynamic_interruptions is None or len(dynamic_interruptions) == 0
-    nostat = next_static_interruption is None
-
-    if nostat and nodyn:
-        trajectory = ode_simulate(dynamics, start_state, timespan, solver=solver)
-        # TODO support event_dim > 0
-        return trajectory, (), timespan[-1], trajectory[..., -1]
-
-    # Leaving these undone for now, just so we don't have to split test coverage. Once we get a better test suite
-    #  for the many possibilities, this can be optimized.
-    # TODO AZ if no dynamic events, just skip the event function pass.
-
-    if dynamic_interruptions is None:
-        dynamic_interruptions = []
-
-    if nostat:
-        # This is required because torchdiffeq.odeint_event appears to just go on and on forever without a terminal
-        #  event.
-        raise ValueError(
-            "No static terminal interruption provided, but about to perform an event sim."
-        )
-    # for linter, because it's not deducing this from the if statement above.
-    assert next_static_interruption is not None
-
```

Comment:
All of this code not torchdiffeq specific moved into internals/dynamical/simulate_to_interruption.py